components/calendar/src/chinese_based.rs

```diff
@@ -309,6 +344,103 @@ impl<C: ChineseBased + CalendarArithmetic> ChineseBasedDateInner<C> {
         debug_assert!(result < MAX_ITERS_FOR_MONTHS_OF_YEAR, "The given year was not a leap year and an unexpected number of iterations occurred searching for a leap month.");
         result
     }
+
+    /// Create a new arithmetic date from a year, month ordinal, and day with bounds checking; returns the
+    /// result of creating this arithmetic date, as well as a ChineseBasedCache - either the one passed in
+    /// optionally as an argument, or a new ChineseBasedCache for the given year, month, and day args.
+    pub(crate) fn new_from_ordinals(
+        year: i32,
+        month: u8,
+        day: u8,
+        cache: &ChineseBasedCache,
+    ) -> Result<ArithmeticDate<C>, CalendarError> {
+        let max_month = Self::months_in_year_cached(cache);
+        if month > max_month {
+            return Err(CalendarError::Overflow {
+                field: "month",
+                max: max_month as usize,
+            });
+        }
+
+        let max_day = Self::days_in_month_cached(month, cache);
+        if day > max_day {
+            return Err(CalendarError::Overflow {
+                field: "day",
+                max: max_day as usize,
+            });
+        }
+
+        // Unchecked can be used because month and day are already checked in this fn
+
+        Ok(ArithmeticDate::<C>::new_unchecked(year, month, day))
+    }
+
+    /// Call `months_in_year_cached` on a `ChineseBasedDateInner`
+    pub(crate) fn months_in_year_inner(&self) -> u8 {
+        Self::months_in_year_cached(&self.1)
+    }
+
+    /// Return the number of months in a given year, which is 13 in a leap year, and 12 in a common year.
+    /// Also takes a `ChineseBasedCache` argument.
+    fn months_in_year_cached(cache: &ChineseBasedCache) -> u8 {
+        if cache.is_leap_year {
+            13
+        } else {
+            12
+        }
+    }
+
+    /// Calls `days_in_month_cached` on an instance of ChineseBasedDateInner
+    pub(crate) fn days_in_month_inner(&self) -> u8 {
+        Self::days_in_month_cached(self.0.month, &self.1)
+    }
+
+    /// Returns the number of days in the given `year`, `month` (see `month_days`);
+    /// this function takes a `ChineseBasedCache` argument.
+    fn days_in_month_cached(month: u8, cache: &ChineseBasedCache) -> u8 {
```

Comment:
Nit: this function isn't really "cached" because it still calls `new_moon` functions.

Suggestion: change this function to `days_in_month(month: u8, new_year: Option<RataDie>)` which computes the new year if not passed as an argument, and then also make `month_days` call this function.